flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/window/WindowOperator.java

```diff
@@ -0,0 +1,711 @@
+
+			// has emitted result for the window
+			if (previousAggResult != null) {
+				// current agg is not equal to the previous emitted, should emit retract
+				if (!equaliser.equalsWithoutHeader(aggResult, previousAggResult)) {
+					reuseOutput.replace((BaseRow) getCurrentKey(), previousAggResult);
+					BaseRowUtil.setRetract(reuseOutput);
+					// send retraction
+					collector.collect(reuseOutput);
+					// send accumulate
+					reuseOutput.replace((BaseRow) getCurrentKey(), aggResult);
+					BaseRowUtil.setAccumulate(reuseOutput);
+					collector.collect(reuseOutput);
+					// update previousState
+					previousState.update(aggResult);
+				}
+				// if the previous agg equals to the current agg, no need to send retract and accumulate
+			}
+			// the first fire for the window, only send accumulate
+			else {
+				// send accumulate
+				reuseOutput.replace((BaseRow) getCurrentKey(), aggResult);
+				BaseRowUtil.setAccumulate(reuseOutput);
+				collector.collect(reuseOutput);
+				// update previousState
+				previousState.update(aggResult);
+			}
+		} else {
+			reuseOutput.replace((BaseRow) getCurrentKey(), aggResult);
+			// no need to set header
+			collector.collect(reuseOutput);
+		}
+	}
+
+	/**
+	 * Registers a timer to cleanup the content of the window.
+	 *
+	 * @param window the window whose state to discard
+	 */
+	private void registerCleanupTimer(W window) {
+		long cleanupTime = cleanupTime(window);
+		if (cleanupTime == Long.MAX_VALUE) {
+			// don't set a GC timer for "end of time"
+			return;
+		}
+
+		if (windowAssigner.isEventTime()) {
+			triggerContext.registerEventTimeTimer(cleanupTime);
+		} else {
+			triggerContext.registerProcessingTimeTimer(cleanupTime);
+		}
+	}
+
+	/**
+	 * Returns the cleanup time for a window, which is
+	 * {@code window.maxTimestamp + allowedLateness}. In
+	 * case this leads to a value greated than {@link Long#MAX_VALUE}
+	 * then a cleanup time of {@link Long#MAX_VALUE} is
+	 * returned.
+	 *
+	 * @param window the window whose cleanup time we are computing.
+	 */
+	private long cleanupTime(W window) {
+		if (windowAssigner.isEventTime()) {
+			long cleanupTime = Math.max(0, window.maxTimestamp() + allowedLateness);
+			return cleanupTime >= window.maxTimestamp() ? cleanupTime : Long.MAX_VALUE;
+		} else {
+			return Math.max(0, window.maxTimestamp());
+		}
+	}
+
+	@SuppressWarnings("unchecked")
+	private K currentKey() {
+		return (K) getCurrentKey();
+	}
+
+	// ------------------------------------------------------------------------------
+
+	/**
+	 * Context of window.
+	 */
+	private class WindowContext implements InternalWindowProcessFunction.Context<K, W> {
+
+		@Override
+		public <S extends State> S getPartitionedState(
+				StateDescriptor<S, ?> stateDescriptor) throws Exception {
+			requireNonNull(stateDescriptor, "The state properties must not be null");
+			try {
+				stateDescriptor.initializeSerializerUnlessSet(WindowOperator.this.getExecutionConfig());
```

Comment:
No need to `initializeSerializerUnlessSet `?
I see `AbstractKeyedStateBackend.getOrCreateKeyedState` will invoke this.